test/unit/math/opencl/device_functions/log1m_exp_test.cpp

```diff
@@ -27,7 +27,7 @@ TEST(MathMatrixCL, log1m_exp) {
   Eigen::VectorXd res = stan::math::from_matrix_cl<-1, 1>(res_cl);
 
   stan::test::expect_near_rel("log1m_exp (OpenCL)", res,
-                              stan::math::log1m_exp(a));
+                              stan::math::log1m_exp(a).eval());
```

Comment:
Guessing this is getting picked up by the more generic version of `expect_near_rel` which then requires the .eval()

```cpp
template <typename T1, typename T2>
void expect_near_rel(const std::string& msg, const T1& x1, const T2& x2,
                     double tol = 1e-8) {
```

I think we could just slap a `require_not_eigen_t<T1, T2>` there? Then it should go to the right specialization

Reply:
First we still need this to match plain Eigen types so your suggestion would not work. Second compiler can not make a conversion of argument and deduce its template at the same time, so this would not work even if we constructed a correct `require_...`

Reply:
> First we still need this to match plain Eigen types so your suggestion would not work.

You can just do the inverse

Changing signature of `expect_near_rel` below passes these tests without `.eval()`. It's better for test writers if they don't have to remember another gotcha in the future so I think it's important to catch handle this stuff in the call

```cpp

// line 54
template <typename EigMat1, typename EigMat2, require_all_eigen_t<EigMat1, EigMat2>...>
void expect_near_rel_finite(const std::string& msg, EigMat1&& x1, EigMat2&& x2) {
  EXPECT_EQ(x1.rows(), x2.rows());
  EXPECT_EQ(x1.cols(), x2.cols());
  auto x1_eval = x1.eval();
  auto x2_eval = x2.eval();
  for (int i = 0; i < x1.size(); ++i)
    expect_near_rel_finite(msg, x1_eval(i), x2_eval(i));
}

template <typename T1, typename T2>
void expect_near_rel_finite(const std::string& msg, const std::vector<T1>& x1,
                            const std::vector<T2>& x2) {
  EXPECT_EQ(x1.size(), x2.size());
  for (size_t i = 0; i < x1.size(); ++i)
    expect_near_rel_finite(x1[i], x2[i]);
}

// line 118
template <typename EigMat1, typename EigMat2, require_all_eigen_t<EigMat1, EigMat2>...>
void expect_near_rel(const std::string& msg, EigMat1&& x1, EigMat2&& x2, double tol = 1e-8) {
  EXPECT_EQ(x1.rows(), x2.rows()) << "expect_near_rel (Eigen::Matrix)"
                                  << " rows must be same size."
                                  << " x1.rows() = " << x1.rows()
                                  << "; x2.rows() = " << x2.rows() << std::endl
                                  << msg << std::endl;
  EXPECT_EQ(x1.cols(), x2.cols())
      << "expect_near_rel:"
      << "cols must be same size."
      << "x1.cols() = " << x1.cols() << "x2.cols() = " << x2.cols() << ")"
      << std::endl
      << msg << std::endl;
  std::string msg2 = "expect_near_rel; require items x1(i) = x2(i): " + msg;
  auto x1_eval = x1.eval();
  auto x2_eval = x2.eval();
  for (int i = 0; i < x1.size(); ++i)
    expect_near_rel(msg2, x1_eval(i), x2_eval(i), tol);
}
```

> Second compiler can not make a conversion of argument and deduce its template at the same time, so this would not work even if we constructed a correct require_.

idt they are simultaneous? template resolution is something like

1. pick all viable signatures candidates (including ones that require implicit conversion)
2. SFINAE the bad ones
3:N(?). Something complex to choose the most similar candidate (and if more than 1 than fail)

There's a chapter on this in the templates book I only skimmed

Reply:
Of course. Using explicit eval() works. Done.